### PR TITLE
Simplify reading of resource files and load intervals in batches

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
- Simplified the code for reading the genes and transcripts files
- Load genes and transcripts in batches of 10K records, since they are A LOT. This code will be similar when loading exons, that are even more. 

Right now a huge amount of records are stored into a list and then saved to database in a single step.
In this branch  the lines from resource files are returned as an iterator. While looping thru the lines, they are parsed and saved in batches of 10K at the time.

### How to test:
- Load genes and transcripts and make sure they are the same number as when loaded using main branch.

### Review:
- [x] Code approved by HS
- [x] Tests executed by CR, GitHub actions
 
This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: src/chanjo2/meta/handle_load_intervals.py [hunk 1]
```diff
@@ -28,15 +29,11 @@
...
-async def parse_resource_lines(url) -> Tuple[List[List[str]], List[str]]:
-    """Returns header and lines of a downloaded resource."""
-    all_lines: List[str] = "".join(
-        [i.decode("utf-8") async for i in stream_resource(url=url)]
-    ).split("\n")
-    all_lines = [line.rstrip("\n") for line in all_lines]
-    resource_header: str = all_lines[0]
-    resource_lines: List[str] = all_lines[1:-2]  # last 2 lines don't contain data
-    return resource_header.split("\t"), resource_lines
+def resource_lines(url) -> Iterator[str]:
```
Comment (northwestwitch) [inline src/chanjo2/meta/handle_load_intervals.py:32]: Simplified this function, that now returns all file lines as a genarator

File: src/chanjo2/meta/handle_load_intervals.py [hunk 2]
```diff
@@ -60,31 +57,39 @@ async def update_genes(build: Builds, session: Session) -> int:
 
     LOG.info(f"Loading gene intervals. Genome build --> {build}")
     url: str = _get_ensembl_resource_url(build=build, interval_type=IntervalType.GENES)
-    header, lines = await parse_resource_lines(url)
 
+    lines: Iterator[str] = resource_lines(url=url)
+
+    header = next(lines).split("\t")
```
Comment (northwestwitch) [inline src/chanjo2/meta/handle_load_intervals.py:63]: First line is the header

**Comment by northwestwitch** [comment]
Test loading genes from this branch:

Genes 37:

image

Genes 38:

image


The numbers are the same as in the original PR: https://github.com/Clinical-Genomics/chanjo2/pull/63#issuecomment-1484778177

File: src/chanjo2/meta/handle_load_intervals.py (hunk 1 above)
Comment (henrikstranneheim) [inline src/chanjo2/meta/handle_load_intervals.py:32]:
```suggestion
def read_resource_lines(url) -> Iterator[str]:
```
What is the url type?

File: src/chanjo2/meta/handle_load_intervals.py [hunk 3]
```diff
@@ -28,15 +29,11 @@
...
-    """Returns header and lines of a downloaded resource."""
-    all_lines: List[str] = "".join(
-        [i.decode("utf-8") async for i in stream_resource(url=url)]
-    ).split("\n")
-    all_lines = [line.rstrip("\n") for line in all_lines]
-    resource_header: str = all_lines[0]
-    resource_lines: List[str] = all_lines[1:-2]  # last 2 lines don't contain data
-    return resource_header.split("\t"), resource_lines
+def resource_lines(url) -> Iterator[str]:
+    """Returns lines of a remote resource file"""
```
Comment (henrikstranneheim) [inline src/chanjo2/meta/handle_load_intervals.py:33]:
```suggestion
    """Returns lines of a remote resource file."""
```

File: src/chanjo2/meta/handle_load_intervals.py [hunk 4]
```diff
@@ -99,35 +108,48 @@ async def update_transcripts(build: Builds, session: Session) -> int:
...
-
-    delete_intervals_for_build(db=session, interval_type=SQLTranscript, build=build)
-    bulk_insert_transcripts(db=session, transcripts=transcripts)
+        try:
+            transcript = TranscriptBase(
+                chromosome=items[0],
+                ensembl_gene_id=items[1],
+                ensembl_id=items[2],
+                start=int(items[3]),
+                stop=int(items[4]),
```
Comment (henrikstranneheim) [inline src/chanjo2/meta/handle_load_intervals.py:133]: Use constant for build

File: src/chanjo2/meta/handle_load_intervals.py [hunk 5]
```diff
@@ -99,35 +108,48 @@ async def update_transcripts(build: Builds, session: Session) -> int:
...
+                refseq_mrna=items[5],
+                refseq_mrna_pred=items[6],
+                refseq_ncrna=items[7],
+                refseq_mane_select=items[8] if build == "GRCh38" else None,
+                refseq_mane_plus_clinical=items[9] if build == "GRCh38" else None,
+                build=build,
```
Comment (henrikstranneheim) [inline src/chanjo2/meta/handle_load_intervals.py:139]:
```suggestion
            if len(transcripts_bulk) > MAX_NR_OF_TRANSCRIPTS:
```

File: tests/src/chanjo2/endpoints/test_intervals.py [hunk 6]
```diff
@@ -190,18 +190,20 @@ def test_load_genes(
...
 
     # GIVEN a number of genes contained in the demo file
-    nr_genes = len(gene_lines[1])
+    nr_genes = len(list(file_handler(path))) - 1
+
     # WHEN sending a request to the load_genes with genome build
     response: Response = client.post(f"{endpoints.LOAD_GENES}{build}")
     # THEN it should return success
-    assert response.status_code == status.HTTP_200_OK
+    # assert response.status_code == status.HTTP_200_OK
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/endpoints/test_intervals.py:205]:
```suggestion
    assert response.status_code == status.HTTP_200_OK
```

**Review by henrikstranneheim — APPROVED** [review]
Much cleaner 💯

**Comment by northwestwitch** [comment]
Test of transcripts loading: 🆗 

Build 37:

image


Build 38:

image


Same numbers as in previous PR: https://github.com/Clinical-Genomics/chanjo2/pull/66#issuecomment-1495629831

File: src/chanjo2/meta/handle_load_intervals.py (hunk 1 above)
Comment (northwestwitch) [reply src/chanjo2/meta/handle_load_intervals.py:32]: str, I'll add it!

File: src/chanjo2/meta/handle_load_intervals.py (hunk 5 above)
Comment (northwestwitch) [reply src/chanjo2/meta/handle_load_intervals.py:142]: Fixed

File: src/chanjo2/meta/handle_load_intervals.py (hunk 4 above)
Comment (northwestwitch) [reply src/chanjo2/meta/handle_load_intervals.py:134]: Fixed

**Pull request merged**